docs/project/how-to-write-an-upgradable-application.md

```diff
@@ -0,0 +1,104 @@
+    {
+        $this->name = [];
+        $this->visible = false;
++       $this->promoted = false;
+    }
+```
+
+you may encounter problems with following code (notice the `promoted` property will not be set)
+
+```php
+namespace App\Model\Product\Parameter;
+
+use Shopsys\FrameworkBundle\Model\Product\Parameter\ParameterData as BaseParameterData;
+
+class ParameterData extends BaseParameterData {
+
+    // ...
+
+    public function __construct()
+    {
+        // parent::__construct() not called intentionally to avoid setting parameter visibility 
+        $this->name = [];
+    }
+```
+
+Adding a simple comment like one above help you identify those places and quickly check them.
+Also, you may be sure this is intentional, and not a bug. 
+
+## Reported bug
+
+Sometimes you may find a bug or request a feature in Shopsys Framework, 
+but until reported (thanks for that!) issue/request is processed, and a new version is released you need to fix it in your code.
+
+Then it's really useful to make a notice in your code, so you can quickly decide in the future whether some parts of your code can be safely deleted as it's already resolved.
+
+```php
+    /**
+     * @param array $matches
+     * @param string|null $locale
+     * @return string
+     */
+    protected function replace(array $matches, ?string $locale): string
+    {
+        // ...
+
+        // $productClassName is required to be set manually until https://github.com/shopsys/shopsys/issues/1693 is resolved
+        $productClassName = 'Shopsys\FrameworkBundle\Model\Product\Product';
+        $imageViews = $this->imageViewFacade->getForEntityIds($productClassName, $this->getIdsForProducts($products));
+        $productActionViews = $this->productActionViewFacade->getForProducts($products);
+
+        // ...
+```
+
+## Deprecations
+
+After upgrade (or before upgrade to another version) you may want to check for deprecations and resolve those reported from the Shopsys namespace.
+It will make your future upgrades easier.
+
+```text
+// example
+User Deprecated: The Shopsys\FrameworkBundle\Component\Image\Config\ImageConfig::setEntityNameResolver() method is deprecated and will be removed in the next major. Use the constructor injection instead.
+```
+
+## Read-model View Objects
+
+When you need to add a new properties to already existing view object, you may find useful to initialize those properties with setters instead of overriding the constructor.
+
+For the future development is better to use getter/setter instead of typed public property.  
+When you try to read from uninitialized property in Twig, you get not much helpful error notice:
+
+```text
+Neither the property "nonSellingPrice" nor one of the methods "nonSellingPrice()", "getnonSellingPrice()"/"isnonSellingPrice()"/"hasnonSellingPrice()" or "__call()" exist
+and have public access in class "App\Model\Product\Detail\ProductDetailView"
+```
+
+Getter is much more verbose and you are immediately aware of the problem:
```

Comment:
I am not sure if I get this section properly - generally, I prefer the original Twig message rather than the "must not be accessed before initialization" message - but you are saying that the later message is displayed thanks to the usage of the getter and it is much more verbose and understandable, is that correct? Maybe we can call each other and brainstorm this section :slightly_smiling_face:

Reply:
reworked